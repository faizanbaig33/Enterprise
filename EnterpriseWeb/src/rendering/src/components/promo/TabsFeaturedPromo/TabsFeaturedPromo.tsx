// Global
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import { useTheme } from 'lib/context/ThemeContext';
import { Field, Text, withDatasourceCheck } from '@sitecore-jss/sitecore-jss-nextjs';
// Components
import { Component } from 'src/helpers/Component';
import { TabsFeaturedPromoTheme } from './TabsFeaturedPromo.theme';
import { ComponentProps } from 'lib/component-props';
import { Tabs } from 'src/helpers/Tabs';
import { Headline } from 'src/helpers/Headline';
import React, { useEffect, useRef, useState } from 'react';
import Promo from './Promo';
import { SvgIcon } from 'src/helpers/SvgIcon';
import classNames from 'classnames';
import { ItemExt } from 'lib/_.Sitecore.Override';
import ToggleSwitch from 'src/helpers/ToggleSwitch/ToggleSwitch';
import { getBreakpoint, useCurrentScreenType } from 'lib/utils/get-screen-type';
import { normalizeGuid } from 'lib/utils/string-utils';

type TabItemsProps = {
  fields: {
    tabItems: Array<Feature.EnterpriseWeb.Elements.Tabs.TabContent>;
    headlineLevel: ItemExt;
    headlineText: Field<string>;
  };
};

type TabCollectionsProps = {
  fields: {
    tabCollections: Array<Feature.EnterpriseWeb.Elements.Tabs.TabCollection & TabItemsProps>;
  };
};

export type TabsFeaturedPromoProps =
  Feature.EnterpriseWeb.Components.Tabs.TabsFeaturedPromo.TabsFeaturedPromo &
    TabCollectionsProps &
    ComponentProps;

const TabsFeaturedPromo = (props: TabsFeaturedPromoProps) => {
  const { fields } = props;
  const { tabCollections } = fields;

  const { themeName, themeData } = useTheme(TabsFeaturedPromoTheme);
  const { currentScreenWidth } = useCurrentScreenType();

  const titlesScroller = useRef<HTMLDivElement>(null);

  //#region initial states and data of tabs

  // ToggleSwitch
  const [activeSwitchIndex, setActiveSwitchIndex] = useState(0);
  //Sets visible tab group
  const [visiblePanelIndex, setVisiblePanelIndex] = useState(0);
  //Sets visible tab panel
  const [currentIndex, setCurrentIndex] = useState(-1);

  //#endregion

  useEffect(() => {
    const selectedId = window.location.hash;

    if (selectedId || fields.defaultActiveTab) {
      for (let collectionIndex = 0; collectionIndex < tabCollections.length; collectionIndex++) {
        let isMatchFound = false;
        for (
          let itemIndex = 0;
          itemIndex < tabCollections[collectionIndex].fields.tabItems.length;
          itemIndex++
        ) {
          if (selectedId) {
            const item = tabCollections[collectionIndex].fields.tabItems[itemIndex];
            const idToCompare = item.fields.contentId.value || `tab-${normalizeGuid(item.id)}`;

            if (selectedId.split('#')[1] === idToCompare) {
              setActiveSwitchIndex(collectionIndex);
              setVisiblePanelIndex(collectionIndex);
              setCurrentIndex(itemIndex);
              isMatchFound = true;
              break;
            }
          } else if (
            tabCollections[collectionIndex].fields.tabItems[itemIndex].fields.contentId.value ===
            (fields.defaultActiveTab as Feature.EnterpriseWeb.Elements.Tabs.TabContent)?.fields
              .contentId.value
          ) {
            setActiveSwitchIndex(collectionIndex);
            setVisiblePanelIndex(collectionIndex);
            setCurrentIndex(itemIndex);
            isMatchFound = true;
            break;
          }
        }
        if (isMatchFound) break;
      }
    } else setCurrentIndex(0);
  }, []);

  // get the tab title appear into view if the tabs are in scrollable section
  useEffect(() => {
    const selectedItem = fields.tabCollections[visiblePanelIndex]?.fields.tabItems[currentIndex];

    const selectedItemId =
      selectedItem?.fields.contentId.value || `tab-${normalizeGuid(selectedItem?.id)}`;

    if (currentIndex >= 0) {
      history.pushState(null, window.location.href, `#${selectedItemId}`);
    }

    if (titlesScroller.current && currentScreenWidth <= getBreakpoint('ml')) {
      // We're using ref of the parent node of the titles in order to avoid bugs which can occur
      // if this component is authored multiple times in single page
      titlesScroller.current
        .querySelector(`[id='tab-${selectedItemId}']`)
        ?.scrollIntoView({ behavior: 'smooth', block: 'nearest', inline: 'center' });
    }
  }, [visiblePanelIndex, currentIndex]);

  const getSwitchTitles = (): string[] => {
    const switchTitles = ['', ''];

    tabCollections.forEach(
      (collection: Feature.EnterpriseWeb.Elements.Tabs.TabCollection, index: number) => {
        switchTitles[index] = (collection?.fields?.mobileToggleTitle as Field<string>).value;
      }
    );

    return switchTitles;
  };

  const selectTab = (itemIndex: number, collectionIndex: number) => {
    setCurrentIndex(itemIndex);
    setVisiblePanelIndex(collectionIndex);
  };

  const renderToggleSwitch = () => {
    if (currentScreenWidth <= getBreakpoint('ml')) {
      return (
        <ToggleSwitch
          switchTitles={getSwitchTitles()}
          activeSwitchIndex={activeSwitchIndex}
          toggleActiveSwtich={() => {
            const _activeSwitchIndex = activeSwitchIndex === 0 ? 1 : 0;
            setVisiblePanelIndex(_activeSwitchIndex);
            setActiveSwitchIndex(_activeSwitchIndex);
            setCurrentIndex(0);
          }}
        />
      );
    }
    return null;
  };

  const renderTitles = (
    collection: Feature.EnterpriseWeb.Elements.Tabs.TabCollection,
    collectionIndex: number
  ) => {
    if (
      (currentScreenWidth <= getBreakpoint('ml') && collectionIndex === visiblePanelIndex) ||
      currentScreenWidth > getBreakpoint('ml')
    ) {
      return (
        <>
          <div className={themeData.classes.tabsClasses.tabHeadlineWrapper}>
            {themeName === 'aw' && (
              <SvgIcon
                icon="orange-triangle"
                className={themeData.classes.tabsClasses.tabHeadlineIcon}
              />
            )}
            <Headline
              fields={collection.fields}
              classes={themeData.classes.tabsClasses.tabHeadline}
            />
          </div>
          <div ref={titlesScroller} className={themeData.classes.tabsClasses.tabTitlesContainer}>
            {currentIndex > 0 && (
              <div
                className={themeData.classes.tabsClasses.leftArrowIcon}
                onClick={() => setCurrentIndex(currentIndex - 1)}
              >
                <SvgIcon icon="arrow-left" />
              </div>
            )}
            <ul className={themeData.classes.tabsClasses.tabTitlesWrapper} key={collectionIndex}>
              {collection.fields.tabItems.map(
                (item: Feature.EnterpriseWeb.Elements.Tabs.TabContent, itemIndex: number) => {
                  return (
                    <li
                      key={itemIndex}
                      id={item.fields.contentId.value || `tab-${normalizeGuid(item.id)}`}
                      className={classNames(
                        themeData.classes.tabsClasses.tabTitle,
                        itemIndex === currentIndex &&
                          collectionIndex === visiblePanelIndex &&
                          themeData.classes.tabsClasses.activeTabTitle
                      )}
                      onClick={() => selectTab(itemIndex, collectionIndex)}
                      onKeyDown={(e) => {
                        if (e.code === 'Enter' || e.code === 'Space') {
                          selectTab(itemIndex, collectionIndex);
                        }
                      }}
                      tabIndex={0}
                      role="button"
                    >
                      <Text tag="h4" field={item.fields.tabTitle as Field<string>} />
                    </li>
                  );
                }
              )}
            </ul>
            {currentIndex < collection.fields.tabItems.length - 1 && (
              <div
                className={themeData.classes.tabsClasses.rightArrowIcon}
                onClick={() => setCurrentIndex(currentIndex + 1)}
              >
                <SvgIcon icon="arrow-right" />
              </div>
            )}
          </div>
        </>
      );
    }

    return null;
  };

  if (!fields) return null;

  const renderTabTitles = () => {
    return (
      <>
        {/* Toggle Switch */}
        {tabCollections.length > 1 && (
          <div className={themeData.classes.tabsClasses.toggleContainer}>
            {renderToggleSwitch()}
          </div>
        )}
        {/* Tab Titles List */}
        {tabCollections.map(
          (
            collection: Feature.EnterpriseWeb.Elements.Tabs.TabCollection,
            collectionIndex: number
          ) => {
            return <>{renderTitles(collection, collectionIndex)}</>;
          }
        )}
      </>
    );
  };

  const renderTabPanels = (): React.ReactElement[] => {
    const panels: Array<React.ReactElement> = [];
    tabCollections.map(
      (collection: Feature.EnterpriseWeb.Elements.Tabs.TabCollection, collectionIndex: number) => {
        return (
          collectionIndex === visiblePanelIndex &&
          collection.fields.tabItems.forEach(
            (tabItem: Feature.EnterpriseWeb.Elements.Tabs.TabContent) => {
              panels.push(<Promo classes={themeData.classes.promoClasses} {...tabItem} />);
            }
          )
        );
      }
    );

    return panels;
  };

  return (
    <Component variant="lg" dataComponent="promo/tabsfeaturedpromo" {...props}>
      <div className="col-span-12">
        <Headline fields={fields} classes={themeData.classes.headline} />
      </div>
      <div className="col-span-12">
        <Tabs classes={themeData.classes.tabsClasses} currentTabIndex={currentIndex}>
          {renderTabTitles()}
          {renderTabPanels()}
        </Tabs>
      </div>
    </Component>
  );
};

export default withDatasourceCheck()<TabsFeaturedPromoProps>(TabsFeaturedPromo);
