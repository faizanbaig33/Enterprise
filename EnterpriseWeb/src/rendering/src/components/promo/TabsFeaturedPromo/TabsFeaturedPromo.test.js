// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
// Local
import TabsFeaturedPromo from './TabsFeaturedPromo';
import defaultData from './TabsFeaturedPromo.mock-data';

it('renders correctly', () => {
  const component = snapshot(TabsFeaturedPromo, { componentProps: defaultData });
  hasDataComponent(component, 'promo/tabsfeaturedpromo');
});
