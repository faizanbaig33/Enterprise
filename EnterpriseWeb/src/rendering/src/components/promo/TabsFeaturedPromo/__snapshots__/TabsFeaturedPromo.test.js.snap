// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  data-component="promo/tabsfeaturedpromo"
  id="id1695242607"
>
  <div
    class="my-8 lg:mx-auto md:max-w-screen-lg theme-white"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 px-m gap-s md:gap-s"
    >
      <div
        class="col-span-12"
      >
        <div
          class="items-top font-sans  text-s ml:text-m font-bold"
        >
          <h2>
            Tabs Featured Promo gray1
          </h2>
          
        </div>
      </div>
      <div
        class="col-span-12"
      >
        <div
          class="flex flex-col ml:flex-row min-h-[542px]"
        >
          <div
            class="w-full ml:w-1/4 ml:pr-s flex flex-col ml:inline-block"
          >
            <div
              class="order-2 max-w-[70%]"
            />
            <div
              class="flex items-center border-t pt-s mb-xxs uppercase"
            >
              <span
                class="mr-xxxs"
              />
              <div
                class="items-top font-sans  text-xs font-heavy"
              >
                <h2>
                  Type of Windows
                </h2>
                
              </div>
            </div>
            <div
              class="relative flex order-3 ml:order-none"
            >
              <ul
                class="order-3 flex ml:inline-block overflow-auto snap-x ml:snap-none"
              >
                <li
                  class="text-body ml:text-xs text-dark-gray cursor-pointer mx-xxs ml:mx-0 py-xxxs ml:px-xxs ml:border-l-2 my-s snap-start whitespace-nowrap !text-secondary font-heavy border-b-2 ml:border-b-0 border-primary"
                  id="tab-5"
                  role="button"
                  tabindex="0"
                >
                  <h4>
                    Tab 5
                  </h4>
                </li>
                <li
                  class="text-body ml:text-xs text-dark-gray cursor-pointer mx-xxs ml:mx-0 py-xxxs ml:px-xxs ml:border-l-2 my-s snap-start whitespace-nowrap"
                  id="tab-6"
                  role="button"
                  tabindex="0"
                >
                  <h4>
                    Tab 6
                  </h4>
                </li>
                <li
                  class="text-body ml:text-xs text-dark-gray cursor-pointer mx-xxs ml:mx-0 py-xxxs ml:px-xxs ml:border-l-2 my-s snap-start whitespace-nowrap"
                  id="tab-7"
                  role="button"
                  tabindex="0"
                >
                  <h4>
                    Tab 7
                  </h4>
                </li>
                <li
                  class="text-body ml:text-xs text-dark-gray cursor-pointer mx-xxs ml:mx-0 py-xxxs ml:px-xxs ml:border-l-2 my-s snap-start whitespace-nowrap"
                  id="tab-8"
                  role="button"
                  tabindex="0"
                >
                  <h4>
                    Tab 8
                  </h4>
                </li>
              </ul>
              <div
                class="absolute left-full top-1/2 -translate-y-1/2 cursor-pointer bg-white before:absolute before:right-full before:inline-block before:h-[30px] before:w-[30px] before:bg-gradient-to-l before:from-white before:content-[""] ml:hidden"
              >
                <span />
              </div>
            </div>
            <div
              class="flex items-center border-t pt-s mb-xxs uppercase"
            >
              <span
                class="mr-xxxs"
              />
              <div
                class="items-top font-sans  text-xs font-heavy"
              >
                <h2>
                  Doors by Type
                </h2>
                
              </div>
            </div>
            <div
              class="relative flex order-3 ml:order-none"
            >
              <ul
                class="order-3 flex ml:inline-block overflow-auto snap-x ml:snap-none"
              >
                <li
                  class="text-body ml:text-xs text-dark-gray cursor-pointer mx-xxs ml:mx-0 py-xxxs ml:px-xxs ml:border-l-2 my-s snap-start whitespace-nowrap"
                  id="tab-1"
                  role="button"
                  tabindex="0"
                >
                  <h4>
                    Tab 1
                  </h4>
                </li>
              </ul>
            </div>
          </div>
          <div
            class="w-full ml:w-3/4"
          >
            <div
              class="theme-gray bg-theme-bg h-full flex flex-col ml:flex-row items-center justify-between p-m ml:px-xl ml:py-l"
            >
              <div
                class="w-full ml:w-3/5 mb-xxs ml:mb-0"
              >
                <div
                  class="relative hidden md:block aspect-video   "
                >
                  <span
                    style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
                  >
                    <span
                      style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 99.83388704318938% 0px 0px 0px;"
                    />
                    <img
                      alt="tabs-featured-promo"
                      data-nimg="responsive"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                    />
                    <noscript />
                  </span>
                </div>
                <div
                  class="relative w-full md:hidden"
                >
                  <span
                    style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
                  >
                    <span
                      style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 99.83388704318938% 0px 0px 0px;"
                    />
                    <img
                      alt="tabs-featured-promo"
                      data-nimg="responsive"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover; object-position: center;"
                    />
                    <noscript />
                  </span>
                </div>
              </div>
              <div
                class="w-full ml:w-2/6"
              >
                <div
                  class="items-top font-sans  text-s ml:text-m text-theme-text font-bold mb-s"
                >
                  <h2>
                    Primary headline
                  </h2>
                  
                </div>
                <div
                  class="font-sans text-sm-s ml:text-s text-theme-text font-medium mb-s"
                >
                  <h4>
                    Optional sub-heading
                  </h4>
                </div>
                <div
                  class="text-body text-dark-gray mb-m"
                >
                  <div
                    class="body-copy"
                  >
                    <div>
                      <p>
                        Est voluptatum tempore nihil provident alias quis enim. Possimus voluptatibus deserunt. Ea sint repellendus nihil eligendi neque. Delectus et veritatis voluptas quia ut.Test1
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  class="mb-s flex items-start md:flex-row md:items-center"
                >
                  <a
                    class="flex w-fit items-center whitespace-nowrap rounded-lg border-4 border-theme-btn-border bg-theme-btn-bg px-m py-[9px] font-sans text-button font-heavy text-theme-btn-text hover:border-theme-btn-border-hover hover:bg-theme-btn-bg-hover hover:text-theme-btn-text-hover disabled:border-gray disabled:text-gray"
                    href="/"
                    role="link"
                    tabindex="0"
                    target="_blank"
                    title=""
                  >
                    Learn More
                    <span
                      class="ml-xxs"
                    />
                    <span
                      class="sr-only"
                    >
                       (Opens in a new tab)
                    </span>
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
